### PR TITLE
Plain text editor implementation based on markdown input

### PR DESCRIPTION
**Pull request by jmartinesp**
## Type of change

- [x] Feature
- [ ] Bugfix
- [ ] Technical
- [ ] Other :

## Content

- Added `MarkdownTextInput`, a `TextInput` counterpart that uses a simple `EditText` under the hood, instead of the RTE library. This will be the default editor unless the user explicitly enables rich text formatting.
- Added `MarkdownTextEditorState` to act as a way to interact with the `EditText` from Compose code.
- Wrapped both `RichTextEditorState` and `MarkdownTextEditorState` into a `TextEditorState` sealed interface with some helper functions.
- Moved the parsing of the current editor state to send a message to `MessageComposerPresenter` from the View, as well as adding mention spans to the plain text editor. Sadly, this means having to add a `MentionSpanProvider` to presenter, which makes it more difficult to test. I also had to add `isTesting` and `showFormatting` properties to the presenter so it can be easily tested. I'll probably iterate on this later.
- Removed the 'enable rich text editor' option from advanced settings, replaced it with a build config option in `MessageComposerConfig` in the `appconfig` module.

Known issues:
- Mentions can't be converted from/to plain text and rich text yet. I'll handle this in another PR.
- We probably want to keep track of whether the text formatting mode was last enabled so we can restore the text editor with that same value when navigating from one room to the other. This will also be handled in another PR.

Needs https://github.com/matrix-org/matrix-rich-text-editor/pull/978.

## Motivation and context

The RTE library has been problematic for some users, most of which didn't want to use rich text at all, so enabling

## Screenshots / GIFs

I made the 2 editors look exactly the same, even replaced the editor used in some previews and the golden screenshots didn't change.

## Tests

Thoroughly test the plain text editor and enabling / disabling text formatting, editing messages, replying to messages, etc.

## Tested devices

- [x] Physical
- [ ] Emulator
- OS version(s): 14

## Checklist

- [ ] Changes have been tested on an Android device or Android emulator with API 23
- [x] UI change has been tested on both light and dark themes
- [ ] Accessibility has been taken into account. See https://github.com/element-hq/element-x-android/blob/develop/CONTRIBUTING.md#accessibility
- [x] Pull request is based on the develop branch
- [x] Pull request includes a new file under ./changelog.d. See https://github.com/element-hq/element-x-android/blob/develop/CONTRIBUTING.md#changelog
- [x] Pull request includes screenshots or videos if containing UI changes
- [ ] Pull request includes a [sign off](https://matrix-org.github.io/synapse/latest/development/contributing_guide.html#sign-off)
- [x] You've made a self review of your PR

### CODE REVIEW

File: features/messages/impl/src/main/kotlin/io/element/android/features/messages/impl/messagecomposer/MessageComposerPresenter.kt [hunk 1]
```diff
@@ -108,12 +112,27 @@ class MessageComposerPresenter @Inject constructor(
...
+
     @OptIn(FlowPreview::class)
     @SuppressLint("UnsafeOptInUsageError")
     @Composable
     override fun present(): MessageComposerState {
         val localCoroutineScope = rememberCoroutineScope()
 
+        // Initially disabled so we don't set focus and text twice
+        var applyFormattingModeChanges by remember { mutableStateOf(false) }
+        val richTextEditorState = richTextEditorStateFactory.create()
```
[the remark this answers is not in the document]
Comment (jmartinesp) [reply features/messages/impl/src/main/kotlin/io/element/android/features/messages/impl/messagecomposer/MessageComposerPresenter.kt:130]: It's actually remembered, the name of this method is misleading. I'll see if we can change it or use some `rememberRIchTextEditorState()` helper function instead.

File: features/messages/impl/src/main/kotlin/io/element/android/features/messages/impl/messagecomposer/MessageComposerPresenter.kt [hunk 2]
```diff
@@ -229,22 +250,68 @@ class MessageComposerPresenter @Inject constructor(
...
+
+        LaunchedEffect(showTextFormatting) {
+            if (!applyFormattingModeChanges) {
+                applyFormattingModeChanges = true
+                return@LaunchedEffect
+            }
+            if (showTextFormatting) {
+                val markdown = markdownTextEditorState.getMessageMarkdown(permalinkBuilder)
+                richTextEditorState.setMarkdown(markdown)
+                // Give some time for the focus of the previous editor to be cleared
```
[the remark this answers is not in the document]
Comment (jmartinesp) [reply features/messages/impl/src/main/kotlin/io/element/android/features/messages/impl/messagecomposer/MessageComposerPresenter.kt:267]: I forgot to remove this comment, nice catch!

**Review by bmarty — COMMENTED** [review]
Thanks, 2 small remarks after a static review, I will test it this morning.

**Review by bmarty — CHANGES_REQUESTED** [review]
Doing some test. The Markdown composer is working as expected, but when in `Rich`, the composer is not emptied when the message is sent, so user can sent the same message several times.

image

**Comment by jmartinesp** [comment]
@bmarty your issue above should be fixed in https://github.com/element-hq/element-x-android/pull/2840/commits/b73a63b95595f3db486c38111595c81bcd30b5a1.

I also added changes to your other comments.

**Review by bmarty — APPROVED** [review]
Thanks, working as expected now!

**Comment by jmartinesp** [comment]
Nice! I'll release a RTE version compatible with these changes and merge the PR.

**Comment by jmartinesp** [comment]
The Maestro job got stuck after finishing so it was in a limbo. I just force merged.

**Pull request merged**